src/main/java/org/cloudfoundry/promregator/scanner/ReactiveAppInstanceScanner.java

```diff
@@ -95,75 +96,97 @@ public void setupMaps() {
 		 */
 		this.applicationMap = new PassiveExpiringMap<>(this.timeoutCacheApplicationLevel, TimeUnit.SECONDS);
 		this.hostnameMap = new PassiveExpiringMap<>(this.timeoutCacheApplicationLevel, TimeUnit.SECONDS);
 		this.domainMap = new PassiveExpiringMap<>(this.timeoutCacheApplicationLevel, TimeUnit.SECONDS);
 	}
-	
-	
+
+
 	public List<Instance> determineInstancesFromTargets(List<Target> targets) {
 		LinkedList<Instance> result = new LinkedList<>();
-		
-		Flux<Target> initialFlux = Flux.fromIterable(targets);
-		
-		Flux<InternalInstance> initialInstancesOnlyApplication = initialFlux.map( target -> {
+
+		Flux<Target> initialFlux = this.expandTargets(Flux.fromIterable(targets));
+
+		Flux<InternalInstance> initialInstancesOnlyApplication = initialFlux.map(target -> {
 			InternalInstance i = new InternalInstance();
-			
+
 			i.target = target;
-			
+
 			i.orgId = getOrgId(i.target.getOrgName());
 			i.spaceId = getSpaceId(i.orgId, i.target.getSpaceName());
 			i.applicationId = getApplicationId(i.orgId, i.spaceId, i.target.getApplicationName());
-			
+
 			return i;
 		});
-		
+
 		Flux<InternalInstance> instancesOfApplications = this.getInstances(initialInstancesOnlyApplication);
-		
+
 		instancesOfApplications.flatMap(instance -> {
 			Mono<String> applUrlMono = this.getApplicationUrl(instance.applicationId, instance.target.getProtocol());
-			
+
 			Mono<String> accessUrlMono = Mono.zip(applUrlMono, Mono.just(instance.target.getPath()))
-			.map(tuple -> {
-				String applUrl = tuple.getT1();
-				if (!applUrl.endsWith("/")) {
-					applUrl += '/';
-				}
-				
-				String path = tuple.getT2();
-				while (path.startsWith("/"))
-					path = path.substring(1);
-				
-				return applUrl + path;
-			});
-			
+					.map(tuple -> {
+						String applUrl = tuple.getT1();
+						if (!applUrl.endsWith("/")) {
+							applUrl += '/';
+						}
+
+						String path = tuple.getT2();
+						while (path.startsWith("/"))
+							path = path.substring(1);
+
+						return applUrl + path;
+					});
+
 			Mono<InternalInstance> newInstance = Mono.zip(Mono.just(instance), accessUrlMono)
-			.map(tuple -> {
-				InternalInstance i = tuple.getT1();
-				i.accessUrl = tuple.getT2();
-				
-				return i;
-			});
-			
+					.map(tuple -> {
+						InternalInstance i = tuple.getT1();
+						i.accessUrl = tuple.getT2();
+
+						return i;
+					});
+
 			return newInstance;
 		}).map(internalInstance -> internalInstance.toInstance()).toIterable().forEach(result::add);
-		
+
 		return result;
 	}
 
+	private Flux<Target> expandTargets(Flux<Target> inputTargets) {
+
+		return inputTargets.flatMap(target -> {
+			if (!target.getApplicationName().equalsIgnoreCase(PromregatorApplication.SPACE_ALL_APPLICATIONS)) {
+				return Mono.just(target);
+			}
+			Mono<String> orgId = getOrgId(target.getOrgName());
+			Mono<String> spaceId = getSpaceId(orgId, target.getSpaceName());
+			Flux<ApplicationResource> apps = allApplicationByOrgAndSpace(orgId, spaceId);
```

Comment:
I am not 100% sure, but if I get you right, then this is more or less a "pre-selection" using an own CF roundtrip. This leads to yet another roundtrip, which is used for resolving the wildcard identifier.

I am little skeptical here, as performance on the CF metadata API is quite crucial from what I saw from the performance tests conducted on promregator. I had tried to avoid as many roundtrips as I could - and (if I get it right) we are adding one. 
Perhaps remember: Scraping intervals are around 10-15s; with a full set of CF API requests, I measured that they may take up to 2s already (with only having two apps in the selection) for discovery only. And the discovery is on the critical path.

I fear that we won't get around to extend https://github.com/promregator/promregator/blob/ea9fc730ce19f8895f10d2213a441b3d47c2a029/src/main/java/org/cloudfoundry/promregator/cfaccessor/ReactiveCFAccessorImpl.java to handle this.

Reply:
This is a relevant point.  It seems we would need a dedicated (target discovery) thread that might use a ReactiveCFAccessorImpl method to refresh the targets with refresh rate hopefully much slower than the metrics update rate  In addition the promregator can expose a endpoint to receive target-update signals (e.g. not receiving the new targets but just a signal that it should recompute its target wild cards)